### PR TITLE
Add icon picker to components and show component's icons

### PR DESCRIPTION
**Pull request by Feranchz**
### What does it do?

Last year we removed FontAwesome dependency on #15133 , on that PR we remove the icons from components on Content Manager and CTB because they were using FontAwesome. With this PR we add again the icon picker on components as well as the icon when you are creating a new entry with a Dynamic Zone, but this time using our own components from the Design System and not FontAwesome

### How to test it?

- **CTB:** Go to any content type and add a new field of type component, you should see a new icon picker where you can select the icon you want
- **CM:** When creating an entry with a dynamic zone, you should see the icon for each component, you can see also the icons at the "Configure the view" page

### CODE REVIEW

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (joshuaellis) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:20]: We have a VisuallyHidden component designed for this, you can do `<VisuallyHidden as="input" />`

File: packages/core/content-type-builder/admin/src/components/ComponentCard/ComponentIcon/ComponentIcon.js
Comment (Boegie19) [inline packages/core/content-type-builder/admin/src/components/ComponentCard/ComponentIcon/ComponentIcon.js:20]: Why does this not link to a Icon form the Icon lib?

File: packages/core/admin/admin/src/content-manager/components/ComponentIcon/ComponentIcon.js
Comment (markkaylor) [inline packages/core/admin/admin/src/content-manager/components/ComponentIcon/ComponentIcon.js:29]: I think it should probably just be in the design system with the other icons

File: packages/core/content-type-builder/admin/src/components/ComponentCard/ComponentIcon/ComponentIcon.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/ComponentCard/ComponentIcon/ComponentIcon.js:20]: Yeah agree and similar comment above with the CM. Since we are already using it in two packages I really think it makes more sense to just have it one place.

**Review by markkaylor — COMMENTED** [review]
- We should check if the icons in color should be in the list
Screenshot 2023-05-12 at 15 48 53
- The empty search state is a bit weird, I think we should display a no result found message or something
Screenshot 2023-05-12 at 15 49 03

File: packages/core/content-type-builder/admin/src/translations/en.json
Comment (markkaylor) [inline packages/core/content-type-builder/admin/src/translations/en.json:202]:
```suggestion
  "IconPicker.search.clear.label": "Clear the icon search"
```

File: api-tests/core/content-type-builder/components.test.api.js
Comment (gu-stav) [inline api-tests/core/content-type-builder/components.test.api.js:71]: Why did you opt for uppercase component icons? Just out of curiosity ...

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:10]: `position` can be used as a prop.

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:13]: Can we use the default theme radius here instead?

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [reply packages/core/content-type-builder/admin/src/components/IconPicker/index.js:20]: Also: I think this pattern is not completely accessible - you should apply the same size as the element underneath and overlay it.

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:35]: Word of caution: when we removed icons we did not migrate the database and let the old keys in there. That means you might receive a string here, referencing an old icon.

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:25]: The label doesn't have any text a screenreader could use to read what this is to a user. Could you please introduce one?

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:72]: What do you need the box for?

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:108]: Needs translation.

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:116]: Needs translation.

**Review by gu-stav — CHANGES_REQUESTED** [review]
It's a good start :) Could you please add tests wherever you can? Since this will be an important part of the component creation I think we should have a good coverage.

**Comment by derrickmehaffy** [comment]
Are these icons optional or required?

**Comment by Boegie19** [comment]
What would happen if you guys deleted one of your icons would that mean if you have that icon it crashes or would it just fall back to the default?

**Comment by derrickmehaffy** [comment]
> > Are these icons optional or required?
> 
> @derrickmehaffy They are optional, if you don't select one it would show the cube icon by default like right now

🥰

File: api-tests/core/content-type-builder/components.test.api.js
Comment (remidej) [reply api-tests/core/content-type-builder/components.test.api.js:71]: I personally don't mind the uppercase if it can avoid maintaining a map or having to do string manipulation

File: packages/core/content-type-builder/admin/src/components/ComponentCard/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/ComponentCard/index.js:106]: Could you explain this change? it seems clickable, so why should it not be a button anymore?

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:21]: Just curious why you decide to go for a blacklist approach and not a whitelist

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:79]: why not `isSelected && 'primary200'`

File: packages/core/content-type-builder/admin/src/components/IconPicker/tests/index.test.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/IconPicker/tests/index.test.js:5]: I don't think fireEvent is needed as you can do all it does can be done with userEvent

File: api-tests/core/content-type-builder/components.test.api.js
Comment (joshuaellis) [reply api-tests/core/content-type-builder/components.test.api.js:71]: ^^^ this, you're create a mental link between the exports of the DS & the DB – could be extremely dangerous, imo create a concrete list of icons you want to be used.

File: packages/core/content-type-builder/admin/src/components/ComponentCard/index.js
Comment (joshuaellis) [reply packages/core/content-type-builder/admin/src/components/ComponentCard/index.js:106]: Is it one of those situations where you click a card and the button is just there to look like a button if they don't click the card? If so, I think its fine - if its not and it does something different we should review this design // implementation.

File: api-tests/core/content-type-builder/components.test.api.js
Comment (gu-stav) [reply api-tests/core/content-type-builder/components.test.api.js:71]: That looks much better to me, thanks! Now the question is: does this still need to be uppercase, or should this use the icon keys from the mapping :)

File: packages/core/content-type-builder/admin/src/components/ComponentCard/ComponentIcon/ComponentIcon.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/ComponentCard/ComponentIcon/ComponentIcon.js:12]: You can make these props on `Wrapper` instead - `borderRadius` too. This way you don't need the styled overwrite anymore 🪄

File: packages/core/content-type-builder/admin/src/components/ComponentCard/ComponentIcon/ComponentIcon.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/ComponentCard/ComponentIcon/ComponentIcon.js:34]: Why are you not setting the `Cube` default value here?

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:19]: This is really nit-picky, but could you please but an empty lines before this import to separate locale from global imports?

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:20]: What do you think about:

1. instead of having this in `utils/` move it into `constants.js`, because it isn't really a utility
2. could you please use a named export? This makes it much easier to grep for those things in the codebase.

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/IconPicker/index.js:44]: Just wondering: do you need the span?

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (joshuaellis) [reply packages/core/content-type-builder/admin/src/components/IconPicker/index.js:19]: We can do an eslint rule for this you know – then just run autofix 👀

Comment (gu-stav) [reply packages/core/content-type-builder/admin/src/components/IconPicker/index.js:20]: That would be amazing.

File: packages/core/content-type-builder/admin/src/components/ComponentCard/index.js
Comment (remidej) [reply packages/core/content-type-builder/admin/src/components/ComponentCard/index.js:106]: thanks for explaining, then assigning to a span seems good to me, but I believe you should remove the `type="button"`

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (remidej) [reply packages/core/content-type-builder/admin/src/components/IconPicker/index.js]: My concern is that if people add new icons to the DS, they probably won't remember to add them to this blacklist, so we could end up with icons that look out of place (like the colored ST/CT icons before)

I would personally more see a whitelist approach since I assume the list of icons we do want here is pretty much definite. I'd like to hear what others think about this though @markkaylor @gu-stav

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (remidej) [reply packages/core/content-type-builder/admin/src/components/IconPicker/index.js:20]: @gu-stav could you share why this would be better? I tend to naturally just ignore the imports when I'm reading/scanning a code file, and splitting imports in 2 blocks makes that harder

File: packages/core/content-type-builder/admin/src/components/ComponentCard/index.js
Comment (gu-stav) [reply packages/core/content-type-builder/admin/src/components/ComponentCard/index.js:106]: Sounds ok to me, but I think you'd have to make sure the `div` behaves like a button then (e.g. adding tabindex, to make it focusable).

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [reply packages/core/content-type-builder/admin/src/components/IconPicker/index.js:20]: @remidej It really comes down to preference. Let's discuss it in the next FE sync and just add an eslint rule for that and we don't have to discuss it ever again :)

File: packages/core/content-type-builder/admin/src/components/IconPicker/index.js
Comment (gu-stav) [reply packages/core/content-type-builder/admin/src/components/IconPicker/index.js]: I think this thread is no longer valid, because @Feranchz made it an explicitly allowlist instead of a blocklist.

**Comment by gu-stav** [comment]
I've checked out the branch today - little piece of feedback

- [ ] I think the delete button should have a tooltip
- [ ] The search looks out of place

Screenshot 2023-05-24 at 11 40 28

- [ ] when I select a component icon, the others start to jump because the outline changes the dimensions
- [ ] I am not sure if it is a good choice to allow adding a trash or drag icon, because it creates a conflict with the actual actions

Screenshot 2023-05-24 at 11 46 39

- [ ] Do we really want to have brand icons in there?

Screenshot 2023-05-24 at 11 48 10

File: packages/core/content-type-builder/server/controllers/validation/__tests__/component.test.js
Comment (gu-stav) [inline packages/core/content-type-builder/server/controllers/validation/__tests__/component.test.js:31]: These should be lowercase too, no? Why do they not fail?

**Review by gu-stav — COMMENTED** [review]
Good progress 🦾

**Review by remidej — COMMENTED** [review]
I reviewed the code and all LGTM 👌 

Regarding QA I think there are a couple of things that can be improved:

- when clicking on a new icon, it creates some layout shift
- when icons are selected in the icon picker, they're not centered in their box
- when opening the search, it would be nice to auto-focus it

https://github.com/strapi/strapi/assets/8087692/ef14da2f-d31f-4857-94da-40c7a017171c

File: packages/core/admin/admin/src/content-manager/pages/EditSettingsView/components/DynamicZoneList.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/pages/EditSettingsView/components/DynamicZoneList.js:58]: If it works great :) I thought `null` would overwrite the default value you set in `defaultProps` :)

File: packages/core/content-type-builder/admin/src/components/IconPicker/tests/index.test.js
Comment (joshuaellis) [reply packages/core/content-type-builder/admin/src/components/IconPicker/tests/index.test.js:32]: Just to counter this, don't use `history` or `Router` instead, just use `MemoryRouter` which is what they tend to advise in the rrd docs 👍🏼

File: packages/core/content-type-builder/server/controllers/validation/__tests__/component.test.js
Comment (gu-stav) [reply packages/core/content-type-builder/server/controllers/validation/__tests__/component.test.js:31]: I agree, but I think it would be good to validate the text is lowercase, because that is basically user-generated text from the schema that can be changed manually and the stricter we are the better :)

File: packages/core/admin/admin/src/content-manager/pages/EditSettingsView/components/DynamicZoneList.js
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/pages/EditSettingsView/components/DynamicZoneList.js:41]:
```suggestion
```

rogue log :)

File: packages/core/content-type-builder/admin/src/components/ComponentCard/index.js
Comment (joshuaellis) [inline packages/core/content-type-builder/admin/src/components/ComponentCard/index.js:99]: How come this needs an id?

**Comment by gu-stav** [comment]
@Feranchz 

> @gu-stav about the searchbar, I'm not following exactly what you mean, should I put more space between the searchbar bottom border and the icons box?

I guess some more space would help. It also looks a bit unbalanced in terms of font-size / proportions, but some space between the search and the icons box might help ...

**Comment by joshuaellis** [comment]
I've added a migration flag, because imo we need to make one to clear out old icons, otherwise there's gonna be a lot of confusion imo.

**Review by gu-stav — COMMENTED** [review]
Code-wise this looks good to me now - I leave the QA up to you. Well done! 👏🏼

I think Josh's comment is a valid concern: there should be a migration path for projects which still have the old icons in their schema files, because if their instance doesn't boot up anymore this is considered a breaking change.

Also: more of a question: is the error thrown at users when their icon doesn't match the regex understandable? I'm wondering if it should say something more human-readable, because the regex took me a while :D

**Comment by gu-stav** [comment]
@Feranchz I think that could be one option. I've mentioned the issue to @nathan-pichon this morning - maybe you two can talk because Nathan is doing a similar migration for review workflows at the moment.

**Comment by nathan-pichon** [comment]
Hello,
We cannot do a migration for all schemas (or at least, I never heard of someone trying such thing).
We can adapt the icon values dynamically in the Strapi instance, but it looks like a global overhead.

I think the app should start (if it doesn't, we should ensure it starts) even if the wrong icon names are filled in those schemas and we should log a warning about that. Maybe displaying a default icon in case there is a problem with the icon stored.

For the validation regex, you can have it required on the creation and not for updates. This way, it will not be a breaking change, but we will still have this "in-between state" were we have old icon names and new ones coexisting (until v5?).

**Comment by gu-stav** [comment]
@Feranchz That would work for me. What I like about Nathan's idea is to log a warning when Strapi boots (e.g. if the icon doesn't match the format) instead of an error. This way you could have a relaxed regex during boot (or none at all) and the strict icon regex when a new icon is set.

Would it make sense to add this before merging this PR?

Also: How do you want to handle documentation (because this PR goes straight into main)? Did you open a PR already or sync with the docs team?

**Review by markkaylor — DISMISSED** [review]
Looks good to me, tested locally.

I tried putting "plop" as an icon and everything still worked.

I think the warning could be nice since users might be confused to see the cube instead of what they have in their schema. But maybe users are already used to that now 🤷‍♂️. I'll let you decide on that @Feranchz

**Comment by gu-stav** [comment]
@Feranchz Your decision :) I think it is fine like it is now and I leave it up to you to decide how far you want to go.

What about the two open questions:

1) Is the regex-warning understandable for a user?
2) How do you want to handle documentation for this feature?

**Review by markkaylor — DISMISSED** [review]
Looks good to me once the documentation is ready 👍

**Review by joshuaellis — APPROVED** [review]

**Comment by joshuaellis** [comment]
Hey @Feranchz, jumping back in – you'll need to remove both those flags, im assuming we've got migration guide & documentation covered? If so, lets push forward. I've solved the reviewing blocker.

**Review by joshuaellis — APPROVED** [review]

**Comment by joshuaellis** [comment]
You can only have one `source` tag :)

**Pull request merged**
